viewer/src/views/threejs/post-processing/ssao.ts

```diff
@@ -0,0 +1,354 @@
+  });
+  return { scene, uniforms };
+}
+
+function createAntialiasScene(diffuseTexture: THREE.Texture): SceneInfo {
+  const uniforms = {
+    tDiffuse: { value: diffuseTexture },
+    resolution: { value: new THREE.Vector2() }
+  };
+  const scene = setupRenderingPass({
+    uniforms,
+    vertexShader: vertexShaderAntialias,
+    fragmentShader: fragmentShaderAntialias
+  });
+  return { scene, uniforms };
+}
+
+const quadCamera = new THREE.OrthographicCamera(-1, 1, 1, -1, 0, 1);
+
+export interface Pass {
+  render: (
+    renderer: THREE.WebGLRenderer,
+    scene: THREE.Scene,
+    camera: THREE.PerspectiveCamera,
+    pass?: SsaoPassType
+  ) => void;
+  setSize: (width: number, height: number) => void;
+  uniforms: Uniforms;
+}
+
+function lerp(value1: number, value2: number, amount: number) {
+  amount = amount < 0 ? 0 : amount;
+  amount = amount > 1 ? 1 : amount;
+  return value1 + (value2 - value1) * amount;
+}
+
+function traverseShaderMaterials(rootNode: THREE.Object3D, callback: (material: THREE.ShaderMaterial) => void) {
+  rootNode.traverseVisible(obj => {
+    // NOTE we cannot rely on instanceof here because THREE is not necessarily the same library
+    if (obj.type !== 'Mesh') {
+      return;
+    }
+    const mesh = obj as THREE.Mesh;
+    if (mesh.material == null || Array.isArray(mesh.material) || mesh.material.type !== 'ShaderMaterial') {
+      return;
+    }
+    const material = mesh.material as THREE.ShaderMaterial;
+
+    callback(material);
+  });
+}
+
+function createKernel() {
+  const kernelSize = 32;
+  const result = [];
+
+  for (let i = 0; i < kernelSize; i++) {
+    const sample = new THREE.Vector3();
+
+    while (sample.length() < 0.5) {
+      // Ensure some distance in samples
+      sample.x = Math.random() * 2 - 1;
+      sample.y = Math.random() * 2 - 1;
+      sample.z = Math.random();
+    }
+
+    sample.normalize();
+
+    let scale = i / kernelSize;
+    scale = lerp(0.1, 1, scale * scale);
+    sample.multiplyScalar(scale);
+
+    result.push(sample);
+  }
+
+  return result;
+}
+
+function createNoiseTexture() {
+  const width = 128;
+  const height = 128;
+
+  const simplex = new SimplexNoise();
+
+  const size = width * height;
+  const data = new Float32Array(size * 4);
+
+  for (let i = 0; i < size; i++) {
+    const stride = i * 4;
+
+    const x = Math.random() * 2 - 1;
+    const y = Math.random() * 2 - 1;
+    const z = 0;
+
+    const noise = simplex.noise3d(x, y, z);
+
+    data[stride] = noise;
+    data[stride + 1] = noise;
+    data[stride + 2] = noise;
+    data[stride + 3] = 1;
+  }
+
+  const result = new THREE.DataTexture(data, width, height, THREE.RGBAFormat, THREE.FloatType);
+  result.wrapS = THREE.RepeatWrapping;
+  result.wrapT = THREE.RepeatWrapping;
+  return result;
+}
+
+export enum SsaoPassType {
+  Regular = 'Regular',
+  Ssao = 'Ssao',
+  SsaoFinal = 'SsaoFinal',
+  Antialias = 'Antialias'
+}
+
+export function createSsaoPass(): Pass {
```

Comment:
Can render targets be reused to avoid allocating so many? Not sure on the memory footprint of this

Reply:
Let us see if it becomes a problem first. It is very easy to mix them up by mistake once we start reusing them. If we go in that direction, we should rather rely on a system like the one postprocessing uses, with a double "buffer". That is, to only have two render targets and swap between them for each effect.